### ISSUE TITLE
feature request: fix overlapping pie radial labels

### ISSUE THREAD

**Issue by knackjax** [plouc/nivo#143]
if the pie is drawn with a bunch of slices that are small, the result is the radial labels are all overlapping as shown in this example

https://www.webpackbin.com/bins/-KylBOM-OLUpXZ8zOhY-

is there a way to fix this? I was thinking maybe varying the horizontal and diagonal lengths between slices might work...

thanks

**Comment by DaveyEdwards** [comment]
Did you ever come up with a solution to this?  Ive just started noticing this and going to try a couple hack fixes.  Your url is broken now so here is a image of the issue for others to see (It gets worse the more small % valued parts there are)


![image](https://user-images.githubusercontent.com/15203899/39828865-62349d04-5371-11e8-865f-e73e67f2d26b.png)

**Comment by knackjax** [comment]
Yes I went in and modified the PieRadialLabels js with an offset variable that gives greater numbers if the height calculated by sin(angle) is a large value.

It is pretty hacky so I didn't bother to contribute back.

linkBase = 5 and linkExponent = 2 seem to work best for my usage but you might have to tweak for your own use

>                 {data.filter(d => skipAngle === 0 || d.angleDegrees > skipAngle).map((d,i) => {
>                     const angle = midAngle(d) - Math.PI / 2
> 
>                     const offset = Math.pow(linkBase, Math.abs(Math.sin(angle)) * linkExponent)
> 
>                     const positionA = positionFromAngle(angle, radius + linkOffset)
>                     const positionB = positionFromAngle(
>                         angle,
>                         radius + linkOffset + linkDiagonalLength + offset
>                     )
>                     let positionC
>                     let labelPosition
>                     let textAnchor
>                     if (angle + Math.PI / 2 < Math.PI) {
>                         positionC = { x: positionB.x + linkHorizontalLength + offset, y: positionB.y }
>                         labelPosition = {
>                             x: positionB.x + linkHorizontalLength + textXOffset + offset,
>                             y: positionB.y,
>                         }
>                         textAnchor = 'start'
>                     } else {
>                         positionC = { x: positionB.x - linkHorizontalLength - offset, y: positionB.y }
>                         labelPosition = {
>                             x: positionB.x - linkHorizontalLength - textXOffset - offset,
>                             y: positionB.y,
>                         }
>                         textAnchor = 'end'
>                     }
![image](https://user-images.githubusercontent.com/20896041/39830082-5f067914-538e-11e8-9fca-8c34d1d7d315.png)

**Comment by DaveyEdwards** [comment]
O wow thanks for the feedback, that is a interesting way to do it.  Something like that could be a good thing to be built in

**Comment by plouc** [comment]
@knackjax nice trick!

**Comment by plouc** [comment]
not that *hacky* :)

**Comment by SuttJ** [comment]
What did you mean by "modified PieRadialLabels js"? I was unable to find that file in @nivo/pie

How would one go about adding this snippet of code to the node_module? I've had no luck trying to use npm link/yarn link to edit locally or finding where exactly to put that code

**Comment by SuttJ** [comment]
Or how would I be able to override the computeRadialLabels() function in nivo-pie.js ?

**Comment by Swoodend** [comment]
Any update on this in terms of a more built in fix?

**Comment by littlecastrum** [comment]
@knackjax would you be willing to give a bit more details on your solution?

**Comment by sibelius** [comment]
![image](https://user-images.githubusercontent.com/2005841/73365423-029dba00-428b-11ea-9213-8cd2f4380beb.png)

I need something similar to RadarChart as well

**Comment by dkim-thomas** [comment]
@plouc @knackjax 

Just wanted to circle back on the implementation provided by @knackjax. Has this solution been implemented into Nivo? Is there a way to add this workaround without altering the core files?

@knackjax would you be able to provide more info on your solution? Seems like there multiple ppl that would like to implement your fix but require a bit more info.

**Comment by SuttJ** [comment]
@dkim-thomas @littlecastrum @Swoodend 

The way I ended up solving this was locally copying the entire `nivo-pie.js` file and importing from that rather than the dependency. 

In this local file, I rewrote the `computeRadialLabels` function. That's what determines the label positions and can handle the overlap.

If anyone is interested, here is the entire file with that modified function: https://gist.github.com/SuttJ/2b21075803906d48ab283998b848d464

Important to note that this was from a while ago

**Comment by vdh** [comment]
Oh FFS, stale bot sucks… closing a huge defect in the pie labels out of the blue like this 😞

**Comment by plouc** [comment]
@vdh please be polite, also if you think that's a "huge defect", contributions are welcome :) automatic label placement in general is quite challenging and this simplistic approach is not uncommon.

Stale bot is actually pretty useful for maintainers to see how important an issue is for users, we cannot cope with every feature request unfortunately.

**Comment by vdh** [comment]
@plouc Apologies for my rude outburst, but Stale bot is quite an unfriendly bot. It's nastiness wrapped in a layer of saccharine.

It's one thing to _say_ contributions are welcome, but the behaviour of Stale makes for an extremely _unwelcoming_ context. It's designed to shut down and suppress issues. It requires users to write unhelpful "+1" noise comments just to shoo it away, which completely defeats the purpose of the reactions that GitHub added for the purpose of reducing noise. It feels dirty, purposefully writing noise comments (that will bother _every_ person subscribed to the issue!) just to keep the reaper bot temporarily at bay. Recreating issues causes more noise (intentionally duplicating issues, making them ripe for "closed for being a duplicate" actioning), and loses the discussion thread from the initial issue.

Stale bot creates a constant, negative pressure to dedicate lots of time and emotional energy to keep issues alive, which is enough of a disincentive to drive earnest discussion and contributors away. I completely understand that there is often never enough time and resources to dedicate towards issues, but this bot is very unwelcoming and discourages contributions.

**Comment by dev-creata** [comment]
Hi @plouc Hi I'm trying to contribute to fix this Issue Can you point me to where I can find the function that computes the path? I found the compute.ts file but making edits to it has no effect on the render (I'm using the storybook)

**Comment by brian-codes** [comment]
@davidkim10 nice solution, I had a look at implementing it, the codebase seems to have diverged a lot since then, differrent files and variablle names, I think this is the equivelant of what you wrote?

https://gist.github.com/brian-codes/2263643770146553118a52d6bdf52528

**Comment by altq33** [comment]
The same problem now on the last version codebase is different, files and var names

**Comment by brian-codes** [comment]
Yes, we've actually had a go at fixing it, one of our team is prepping a PR.
